### PR TITLE
Just opening for comments

### PR DESCRIPTION
**Pull request by nickjalbert**
No need to merge @andyk (unless you want to), just wanted to comment on DVN.

### CODE REVIEW

File: contrib/dvn.py [hunk 1]
```diff
@@ -0,0 +1,203 @@
...
+    def _restore(self, tmp_checkpoint_dir):
+        self.v_model = load_model(os.path.join(tmp_checkpoint_dir, "/model"))
+
+    def get_action(self, episode_num, state):
+        prob_random_action = min(1., (self.params["num_init_random_actions"] + self.params["epsilon"]) / (episode_num + 1.))
+        self.mlflow_log_metric("prob random action", prob_random_action, step=episode_num)
+        if random.random() < prob_random_action:
+            action = self.env.action_space.sample()
+        else:
+            state_vals = tf.concat([self.v_model(np.asarray(self.env.get_afterstate(state, a)[0])[np.newaxis])
```
Comment (nickjalbert) [inline contrib/dvn.py]: Should this also be getting the canonical board as well as the afterstate?

File: contrib/dvn.py [hunk 2]
```diff
@@ -0,0 +1,203 @@
...
+                    vals = self.v_model(afterstates)
+                    next_ca = []  # next_canonical_afterstates
+                    for n_a in range(self.env.action_space.n):  # n_a is next_action
+                        next_as = [self.env.get_afterstate(s, n_a)[0] for s in next_states]
+                        next_canonicals = np.asarray([self.env.get_canonical_board(s) for s in next_as]).reshape(next_states.shape)
+                        next_ca.append(self.v_model(next_canonicals))
+                    next_vals_all = tf.concat(next_ca, 1)
+                    next_vals = tf.expand_dims(tf.reduce_max(next_vals_all, axis=1), 1)
+                    disc_next_v = (1 - dones) * self.params["gamma_ie_discount_rate"] * next_vals
+                    td_target = rewards + disc_next_v
```
Comment (nickjalbert) [inline contrib/dvn.py]: If I understand correctly, you are calculating the `td_target` by adding the reward of the transition from `S` to `S'` to the max discounted value for all possible canonical afterstates of `S'`.  Should there also be a term for "immediate" reward of transitioning from `S'` to its canonical afterstate?

To make it concrete, let's say board A is:

```
-   -   -   -
-   -   -   -
16  16  -   -
32  -   -   -
```
we decide to go LEFT so we get the following afterstate (board B) with 32 reward: 

```
-   -   -   -
-   -   -   -
32  -   -   -
32  -   -   -
```
and then we end up with board C:

```
-   -   -   -
-   -   2   -
32  -   -   -
32  -   -   -
```

Let's assume all those boards happen to be in their canonical rotation.  The memory tuple would be something like ```(board_A, LEFT, board_B, 32, 0, board_C)```.

So then, when we're training:

``` vals = self.v_model(board_B)```

and `next_vals` is the max (with respect to the `self.v_model` value) of all the possible canonical afterstates of board C.  The thing that strikes me is that only the `self.v_model` value of C's afterstates and not also the immediate reward of transitioning from C to the canonical afterstate are considered when figuring out our `next_val`.

To go back to the example, when looking at the canonical afterstate of C and DOWN, we get the board:
```
-   -   -   -
-   -   -   -
-   -   -   -
64  -   2   -
```

We also get 64 reward to transition to that afterstate that doesn't figure into the `next_val` (and hence into the TD target).

Comment (nickjalbert) [reply contrib/dvn.py]: I typed that, and I think I changed my mind on it.  Mainly because once you take the action DOWN from board C, then you'll credit the afterstate

```
-   -   -   -
-   -   -   -
-   -   -   -
64  -   2   -
```
 with the 64 reward and then I think things will basically work out (e.g. next time you get to C you will check the `self.v_model` for DOWN and it should know about the 64 immediate reward.

Comment (nickjalbert) [reply contrib/dvn.py]: OK last comment on this I swear!

The value function V<sub>&pi;</sub>(S) is usually the discounted expected return of policy &pi; from state S.

What I think you're learning here is the discounted expected return of your policy from canonical afterstate S AND the value of the immediate reward of getting into that canonical afterstate S (e.g. the 64 reward we got for the transition between board C and board D above get credited to board D). 

If the above is correct, you can imagine this could hinder learning because there might be a large number of ways to get to board D, some with high immediate reward and some with no immediate reward, and this will confuse the your value estimate.

File: contrib/dvn.py [hunk 3]
```diff
@@ -0,0 +1,203 @@
...
+                    val_targets = (1 - alpha) * vals + alpha * td_target
+                    loss = self.loss_fn(vals, val_targets)
```
Comment (nickjalbert) [inline contrib/dvn.py]: This is probably just me not understanding `GradientTape`, but how does the optimizer know which inputs to the NN (which, I believe, are canonical afterstates) are associated with which `vals`/`val_targets`?

File: contrib/dvn.py (hunk 2 above)
Comment (nickjalbert) [reply contrib/dvn.py]: ### Standard Q-learning
Choose A from S &epsilon;-greedily
Take action A, observe R and S'
Q(S,A) &#8592; Q(S,A) + &alpha;[R + &gamma;max<sub>a</sub>Q(S',a) - Q(S,A)]
S &#8592; S'

### 2048 "model-lite" V-learning

Let V(S) be the estimated discounted expected reward in state S
Let I(S,A) be the immediate reward for from taking action A in state S (i.e. one step reward)
Let C(S,A) be the canonical afterstate one arrives at by taking action A in state S






Choose A from S &epsilon;-greedily, where greedy is max<sub>A</sub>[I(S,A) + V(C(S,A))]
Take action A, observe R [which is guaranteed to be I(S,A)] and S'
V(C(S,A))  &#8592; V(C(S,A)) + &alpha;[&gamma;max<sub>A</sub>[I(S',A) + V(C(S',A))] - V(C(S,A))]
